keps/349-avoid-coscheduling/README.md

```diff
@@ -0,0 +1,418 @@
+and make progress.
+-->
+
+## Proposal
+
+We introduce a mechanism to avoid concurrent scheduling of Job pods. More
+precisely, we block admission of new workloads until the first batch of the
+Job is scheduled. Additionally, we await with starting the Job until all other
+admitted workloads have their pods scheduled. This behavior can be opted-in by
+a Job by a new annotation.
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation. What is the desired outcome and how do we measure success?.
+The "Design Details" section below is for the real
+nitty-gritty.
+-->
+
+### User Stories (Optional)
+
+<!--
+Detail the things that people will be able to do if this KEP is implemented.
+Include as much detail as possible so that people can understand the "how" of
+the system. The goal here is to make this feel real for users without getting
+bogged down.
+-->
+
+#### Story 1
+
+As a machine learning researcher I want to run a workload, comprising tens of
+pods which need to all run at the same time. Autoscaling the cluster and
+scheduling of all pods in this workload take a couple of minutes and I want to
+have a guarantee that other workloads scheduled within this time interval would
+not delay execution of my workload.
+
+My use can be supported by adding `kueue.x-k8s.io/wait-for-pods-scheduled`
+annotation to the jobs I run.
+
+#### Story 2
+
+As an administrator of a batch processing in my organization I want to be able
+to configure kueue quotas for on-demand resources a little higher than what is
+available in the quiet period. However, I don't want to risk two workloads
+deadlock in case of issues with provisioning nodes to satisfy the configured
+quota.
+
+My use can be supported by adding `kueue.x-k8s.io/wait-for-pods-scheduled`
+annotation to all jobs.
+
+### Notes/Contraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+### Risks and Mitigations
+
+If the workload fails to schedule its pods it could block admission of other
+workloads indefinitely.
+
+The risk can be mitigated by using the Job's `ActiveDeadlineSeconds` timeout
+to terminate a started job, which results in finishing the associated workload.
+
+Still the mitigation might be problematic as the timeout needs to include both
+the scheduling and execution time for a job, thus needs to be higher than one
+would want.
+
+Thus, we consider introduction of timeout on reaching the `PodsScheduled`
+condition by a workload since its job start (see:
+[Timeout on reaching the PodsScheduled condition](#timeout-on-reaching-the-podsscheduled-condition))
+as a possible future improvement.
```

Comment:
I think we need this now, we can't block forever.

Reply:
I see, moving it out from Alternatives